### PR TITLE
290 tests v2

### LINKED ISSUE #290: Make running tests easier
Running tests involves a lot of friction this makes the dev cycle slower. Here are some ideas,

## 1. Dockerize the tests
A priviledged Alpine/Debian container to run the tests.
1. One time setup (the image and container)
```sh
# Build the image
docker build -t starknet-hardhat-plugin-tester .

# Uncomment the line below to remove the old container
# docker rm starknet-hardhat-plugin-tester

# Create a new container
docker create --name starknet-hardhat-plugin-tester starknet-hardhat-plugin-tester
```

2. Running tests.
```sh
# Copy the latest plugin
docker cp . starknet-hardhat-plugin-tester:/node

# The new image will have the files
docker run -it starknet-hardhat-plugin-tester /bin/sh
```

3. Then inside container we just do `npm i` and `npm run test-general-tests` etc. to run tests.

## 2. Mock devnet etc dependencies

> We might have discussed this before and decided to go the other way, let me know what you guys think.

An Axios mock would be really neat. This would allows tests to run much faster and be great for TDD.
So for example a test for devnet.mint #285 would just check if the `mint` method makes a request to `/mint` endpoint with the correct data.
The drawback would be if devnet changes the way request to mint tokens is handled the plugin might be broken while the tests still pass. However with devnet instance locked to a specific version we should be all good.

### PR DESCRIPTION
**Pull request by shramee**
## Usage related changes

Resolves #290

## Development related changes

- Tests can be run in a docker container
- Individual tests can be run with `npm run test-general-tests -- account-test`

## Checklist:

-   [x] Formatted the code
-   [x] No linter errors + tried to avoid introducing linter warnings
-   [x] Performed a self-review of the code
-   [x] Rebased to the last commit of the target branch (or merged it into my branch)
-   [x] Documented the changes
-   [ ] Updated the `test` directory (with a test case consisting of `network.json`, `hardhat.config.ts`, `check.ts`)
-   [x] Linked issues which this PR resolves
-   [ ] Created a PR to the `plugin` branch of [`starknet-hardhat-example`](https://github.com/Shard-Labs/starknet-hardhat-example):
    -   < EXAMPLE_REPO_PR_URL > 
    -   [ ] Modified `test.sh` to use my example repo branch
    -   [ ] Restored `test.sh` to to use the original branch (after the example repo PR has been merged)
-   [ ] All tests are passing (for external contributors who don't have access to the CI/CD pipeline)

### CODE REVIEW

**Comment by shramee** [comment]
@FabijanC Here's the new PR for #296.

**Review by FabijanC — CHANGES_REQUESTED** [review]
There are failing tests, perhaps `/restart` is being invoked on a wrong network? I suggest you check the circleci logs.

Also, can you update the list of failing tests (or delete it?) in the PR description?

All in all, I am not very prone to accepting this PR as it currently is. I still don't fully understand what's going on and would have to invest more time in reviewing it. I cannot accept it on superficial understanding.

But more than that, I don't like that, if I understood correctly, what this PR is basically introducing is modifying the plugin to do:
```js
if  (testing) { do_this(); } else { do_that(); }
```

I am ok with modifying the testing code (the plugin tests), but not the code being tested (the plugin source code). Because that way we are no longer testing the actual code users will be using.

Let me know if I misunderstood something.

File: scripts/devnet-run.sh [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+	container_id=$(docker run --name starknet_hardhat_dind_devnet -d -p 0.0.0.0:$PORT:$PORT shardlabs/starknet-devnet --seed 42)
+	echo "Running devnet in container starknet_hardhat_dind_devnet $container_id"
+	HOST=host.docker.internal
+
+else
+	starknet-devnet --host $HOST --port $PORT --seed 42 >/dev/null 2>&1 &
+	echo "Spawned devnet with PID $!"
+fi
+
+echo "Waiting for devnet... "
```
[the remark this answers is not in the document]
Comment (shramee) [reply scripts/devnet-run.sh:30]: Aye, but that depends on the curl call failing. However with ports exposed it doesn't fail but just returns an empty response (which means that it's not live yet).
So added some code to continuously check until the response is not empty. And also logs number of try each second... (like a seconds counter)

[the remark this answers is not in the document]
Comment (FabijanC) [reply scripts/devnet-run.sh:30]: Why is it bad that it depends on the curl call failing? Not failing gracefully? I think the curl command can also log the retries it's making.

[the remark this answers is not in the document]
Comment (shramee) [reply scripts/devnet-run.sh:30]: Ohh we need the disgrace for it to keep re-trying, it should fail like the sky falling...
But it fails without an error with an empty response which is the problem...

[the remark this answers is not in the document]
Comment (FabijanC) [reply scripts/devnet-run.sh:30]: And surely no flags/options can be passed to the command to fix it?

[the remark this answers is not in the document]
Comment (shramee) [reply scripts/devnet-run.sh:30]: [Retry on any and all errors](https://everything.curl.dev/usingcurl/downloads/retry#retry-on-any-and-all-errors) sounds promising, but I quite like the stopwatch kinda thing. I think it improves the testing experience...
But let me know if we should see if retry on all errs works for us and replace with that.

**Comment by shramee** [comment]
Hi Fabijan,
I think there might be another way to go about this. Without having to make any changes to the code itself. We should be able to use a lot of stuff from here.
I will update you in the first week in April, is that cool?

**Comment by FabijanC** [comment]
> Hi Fabijan, I think there might be another way to go about this. Without having to make any changes to the code itself. We should be able to use a lot of stuff from here. I will update you in the first week in April, is that cool?

Hi Shramee, thanks for replying. That sounds good!

**Comment by shramee** [comment]
Closed in favor of #343

**Pull request closed without merging**